File: packages/core/test/routing-corpus.test.ts
/**
 * Replays test/fixtures/routing-corpus.txt, the ordinary ways each question gets typed. The
 * JetBrains plugin replays the same file, so a phrasing fixed here cannot regress there.
 */
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { describe, expect, it } from 'vitest';
import {
  isDatabaseOverviewQuestion,
  isMetadataQuestion,
  isSchemaAdviceQuestion,
  isWriteRequest,
} from '../src/schema-match.js';
import { isCapabilityQuestion } from '../src/scope.js';

const FIXTURE = join(dirname(fileURLToPath(import.meta.url)), 'fixtures', 'routing-corpus.txt');

function loadCorpus(): readonly (readonly [string, string])[] {
  return readFileSync(FIXTURE, 'utf8')
    .split('\n')
    .filter((line) => line.trim() !== '' && !line.startsWith('#'))
    .map((line) => {
      const tab = line.indexOf('\t');
      if (tab < 0) throw new Error(`malformed corpus line: ${line}`);
      return [line.slice(0, tab), line.slice(tab + 1)] as const;
    });
}

/** The engine's own order of checks, so the corpus measures routing as it actually happens. */
function routeOf(question: string): string {
  if (isCapabilityQuestion(question)) return 'capability';
  if (isWriteRequest(question)) return 'write';
  if (isSchemaAdviceQuestion(question) || isDatabaseOverviewQuestion(question)) return 'advice';
  return isMetadataQuestion(question) ? 'listing' : 'data';
}

const CORPUS = loadCorpus();

describe('routing corpus', () => {
  it('has thousands of questions rather than a sample', () => {
    expect(CORPUS.length).toBeGreaterThanOrEqual(4000);
  });

  it('routes every question in the corpus', () => {
    const misrouted: string[] = [];
    for (const [expected, question] of CORPUS) {
      const actual = routeOf(question);
      if (actual === expected) continue;
      // 'listing' is a hint for the repair loop rather than a separate path: both still
      // generate SQL, so a data question read as a listing is not a misroute.
      if (expected === 'data' && actual === 'listing') continue;
      misrouted.push(`${expected} -> ${actual}: ${question}`);
    }
    const summary = `${misrouted.length}/${CORPUS.length} misrouted\n${misrouted.slice(0, 25).join('\n')}`;
    expect(misrouted, summary).toEqual([]);
  });

  it('covers every route', () => {
    expect([...new Set(CORPUS.map(([route]) => route))].sort()).toEqual([
      'advice',
      'capability',
      'data',
      'listing',
      'write',
    ]);
  });
});

File: packages/jetbrains/src/test/kotlin/com/rahulmahadik/asksql/ide/engine/RoutingCorpusTest.kt
package com.rahulmahadik.asksql.ide.engine

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

/**
 * Replays the same routing corpus the core suite does, from the same file. Identical regex text is
 * not identical behaviour - the two engines differ on flags, escaping and word boundaries - so the
 * only check worth having is that both route thousands of real phrasings the same way.
 */
class RoutingCorpusTest {

    private fun corpusFile(): File = listOf(
        File("../core/test/fixtures/routing-corpus.txt"),
        File("../../packages/core/test/fixtures/routing-corpus.txt"),
        File(System.getProperty("user.dir"), "../core/test/fixtures/routing-corpus.txt"),
    ).firstOrNull { it.exists() }
        ?: error("routing-corpus.txt not found - regenerate it with `node tools/gen-routing-corpus.mjs` in packages/core")

    private fun load(): List<Pair<String, String>> = corpusFile()
        .readLines()
        .filter { it.isNotBlank() && !it.startsWith("#") }
        .map { line ->
            val tab = line.indexOf('\t')
            require(tab >= 0) { "malformed corpus line: $line" }
            line.substring(0, tab) to line.substring(tab + 1)
        }

    /** The pipeline's own order of checks, so the corpus measures routing as it actually happens. */
    private fun routeOf(question: String): String = when {
        Scope.isCapabilityQuestion(question) -> "capability"
        EnginePipeline.isWriteRequest(question) -> "write"
        EnginePipeline.isSchemaAdviceQuestion(question) ||
            EnginePipeline.isDatabaseOverviewQuestion(question) ||
            EnginePipeline.isRelationshipQuestion(question) -> "advice"
        EnginePipeline.isMetadataQuestion(question) -> "listing"
        else -> "data"
    }

    @Test
    fun `every question in the corpus routes the way core routes it`() {
        val corpus = load()
        assertTrue("corpus is too small to be a corpus: ${corpus.size}", corpus.size >= 4000)
        assertEquals(
            "the corpus lost a route",
            setOf("advice", "capability", "data", "listing", "write"),
            corpus.map { it.first }.toSet(),
        )

        val misrouted = corpus.mapNotNull { (expected, question) ->
            val actual = routeOf(question)
            // A data question read as a listing still generates SQL, so it is not a misroute.
            if (actual == expected || (expected == "data" && actual == "listing")) null
            else "$expected -> $actual: $question"
        }

        assertEquals(
            "${misrouted.size}/${corpus.size} misrouted\n" + misrouted.take(25).joinToString("\n"),
            emptyList<String>(),
            misrouted,
        )
    }
}
